akka-stream-tests/src/test/scala/akka/stream/scaladsl/RetryFlowSpec.scala

```diff
@@ -0,0 +1,572 @@
+          minBackoff = 10.millis,
+          maxBackoff = 5.seconds,
+          randomFactor = 0d,
+          maxRetries = 3,
+          flow)(decideRetry = {
+          case ((_, _), (Success(i), ctx)) if i < 5 => Some((i + 1) -> (ctx + 1))
+          case _                                    => None
+        })
+      //#retry-success
+
+      val (source, sink) = TestSource.probe[Int].map(_ -> 0).via(retryFlow).toMat(TestSink.probe)(Keep.both).run()
+
+      sink.request(4)
+
+      source.sendNext(5)
+      sink.expectNext(Success(5) -> 0)
+
+      source.sendNext(2)
+      sink.expectNext(Success(5) -> 3)
+
+      source.sendComplete()
+      sink.expectComplete()
+    }
+
+  }
+
+  "Backing off" should {
+    "have min backoff" in {
+      val minBackoff = 200.millis
+      val (source, sink) = TestSource
+        .probe[(Int, Int)]
+        .via(RetryFlow.withBackoffAndContext(minBackoff, 5.second, 0d, 3, failEvenValuesFlow)(incrementFailedValues))
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1 -> 0)
+      sink.expectNext(80.millis, Success(1) -> 0)
+
+      source.sendNext(2 -> 0)
+      sink.expectNoMessage(minBackoff)
+      sink.expectNext(Success(3) -> 1)
+    }
+
+    "use min backoff for every try" in {
+      val minBackoff = 50.millis
+      val maxRetries = 3
+      val (source, sink) = TestSource
+        .probe[(Int, Int)]
+        .via(RetryFlow.withBackoffAndContext(minBackoff, 5.seconds, 0d, maxRetries, failAllValuesFlow)(
+          incrementFailedValues))
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(1)
+
+      source.sendNext(10 -> 0)
+      sink.expectNoMessage(minBackoff * maxRetries)
+      sink.expectNext(FailedElem -> maxRetries)
+    }
+
+    "exponentially backoff between retries" in {
+      val NumRetries = 7
+
+      val nanoTimeOffset = System.nanoTime()
+      case class State(retriedAt: List[Long])
+
+      val flow = FlowWithContext.fromTuples[State, NotUsed, State, NotUsed, NotUsed](Flow.fromFunction {
+        case (State(retriedAt), _) => State(System.nanoTime - nanoTimeOffset :: retriedAt) -> NotUsed
+      })
+
+      val (source, sink) = TestSource
+        .probe[(State, NotUsed)]
+        .via(RetryFlow.withBackoffAndContext(10.millis, 5.seconds, 0d, NumRetries, flow) {
+          case (_, (s, _)) => Some(s -> NotUsed)
+          case _           => None
+        })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(1)
+
+      source.sendNext(State(Nil) -> NotUsed)
+      val (State(retriedAt), _) = sink.expectNext()
+
+      val timesBetweenRetries = retriedAt
+        .sliding(2)
+        .collect {
+          case before :: after :: Nil => before - after
+        }
+        .toIndexedSeq
+
+      timesBetweenRetries.reverse should strictlyIncrease
+
+      source.sendComplete()
+      sink.expectComplete()
+    }
+
+  }
+
+  "Aborting" should {
```

Comment:
Matter of taste but I'd keep these test cases under the same main "should" header, because it is all about how "RetryFlow.withBackoff" behaves.